lib/code_extractor.rb

```diff
@@ -0,0 +1,126 @@
+require 'yaml'
+
+# Class to extract files and folders from a git repository, while maintaining
+# The git history.
+module CodeExtractor
+  def run
+    Runner.new.extract
+  end
+  module_function :run
+
+  class Config
+    def initialize(config_file = 'extractions.yml')
+      @config = YAML.load_file(config_file)
+
+      @config[:destination]       = File.expand_path(@config[:destination])
+      @config[:upstream_branch] ||= "master"
+
+      validate!
+    end
+
+    def [](key)
+      @config[key]
+    end
+
+    def inspect
+      @config.inspect
+    end
+    alias to_s inspect
+
+    def validate!
+      missing = %i[name destination upstream upstream_name extractions].reject { |k| @config[k] }
+      raise ArgumentError, "#{missing.map(&:inspect).join(", ")} key(s) missing" if missing.any?
+    end
+  end
+
+  class GitProject
+    attr_reader :name, :url, :git_dir, :new_branch, :source_branch
+
+    def initialize name, url
+      @name   = name
+      @url    = url
+    end
+
+    def clone_to destination, origin_name = "upstream"
+      @git_dir ||= destination
+
+      if Dir.exist?(git_dir)
+        raise "Not a git dir!" unless system "git -C #{git_dir} status"
+      else
+        puts 'Cloning…'
```

Comment:
Are the unicode ellipses intentional? `…` as opposed to `...`

Reply:
Yes, but I was coping what was being used already in the project.